Data/Queue/Base.juvix

```diff
@@ -0,0 +1,100 @@
+--- This module provides an implementation of a First-In, First-Out (FIFO)
+--- queue based on Okasaki's "Purely Functional Data Structures." Ch.3.
+---
+--- This Okasaki Queue version data structure ensures worst-case constant-time
+--- performance for basic operations such as push, pop, and front.
+---
+--- The OkasakiQueue consists of two lists (front and back) to separate
+--- the concerns of adding and removing elements for ensuring efficient
+--- performance.
+module Data.Queue.Base;
+
+open import Stdlib.Prelude;
+
+--- A first-in-first-out data structure
+type Queue (A : Type) :=
+  | queue : List A -> List A -> Queue A;
+
+--- 𝒪(1). The empty ;Queue;.
+empty : {A : Type} -> Queue A;
+empty := queue nil nil;
+
+--- 𝒪(1). Returns ;true; when the ;Queue; has no elements.
+isEmpty : {A : Type} -> Queue A -> Bool;
+isEmpty (queue nil nil) := true;
+isEmpty _ := false;
+
+--- 𝒪(1). Returns first element of the ;Queue;, if any.
+head : {A : Type} -> Queue A -> Maybe A;
+head (queue nil _) := nothing;
+head (queue (x :: _) _) := just x;
+
+--- 𝒪(1). Removes the first element from the ;Queue;. If the ;Queue; is empty then returns ;nothing;.
+tail : {A : Type} -> Queue A -> Maybe (Queue A);
+tail (queue nil _) := nothing;
+tail (queue (_ :: front) back) := just (queue front back);
+
+--- 𝒪(n) in the worst-case scenario, but 𝒪(1) amortized 
+--- due to infrequentcalls. 
```

Comment:
```suggestion
```

The reason it is 𝒪(1) amortized can't be because of infrequent calls. The amount of times a function is called does not affect its asymptotic behaviour. I suspect the reason why this is amortized is because the first clause is hit much less often compared to the second clause.